doc/design/layer_generation/00.how_to_write_a_layer.md

```diff
@@ -0,0 +1,249 @@
+  
+  optional Attribute default_value = 4; // default value.
+  optional Attribute max_value = 5;    // max value.
+  optional Attribute min_value = 6;   // min value.
+}
+
+// Argument Define the Supported InputTypes.
+message ArgumentDef {
+   	// Supported Input Type.
+   	// The data type of input/output.
+   	repeated DataType data_type = 1; 
+   	// 0 means it is not a sequence. 1 means a plain sequence. 2 means a nested sequence.  One layer could support many sequence type.
+   	repeated uint32 seq_nested_level = 2;
+    	
+   	// In paddle, some layer can handle variable length input.
+   	// If some input is repeatable, it means there are one or many inputs as the same input type.
+   	required bool repeatable = 3;
+    	
+	// In Paddle, a layer could return many outputs. Each output contains a different name.
+   	required string name = 4;
+   	
+   	// Comments
+  	required string description = 5;
+}
+
+message LayerDef {
+    required string type = 1;  // Layer type, such as 'fc', 'conv'
+    required string description = 2;  // Layer description & comments.
+    
+    
+    repeated ArgumentDef inputs = 3;
+    
+    
+    message ParameterDef {
+        repeated AttributeDef attributes = 1;  // Parameter Attributes Definition.
+    }
+    
+    // Each input of Paddle Layer should contain zero or one parameter.
+    // so parameter_attr.size() == inputs.size()
+    repeated ParameterDef parameter_attr = 5;
+    
+    // Set the bias attribute, If this layer support bias.
+    optional ParameterDef bias_attr = 6;
+    
+    // The Layer Attributes.
+    repeated AttributeDef layer_attr = 7;
+}
+
+// Define the layer's output types by given input types.
+message LayerOutputType {
+	// Output name, Each Paddle Layer could have multiple outputs.
+	optional string name = 1;
+	
+	// Output type
+	required DataType type = 2;
+	required uint32 size = 3;
+	required uint32 seq_nested_level = 4;
+	
+}
+```
+
+### C++ 端暴露LayerDef/LayerOutputType Protobuf.
+
+基本想法:
+
+* 对于每一种类型的Layer，Paddle根据Layer的名字约定两个全局函数的名字。例如，对于FC Layer，全局函数的名字是 `__get_fc_layer_definition__` 和 `__get_fc_layer_output_type__`。 这两个全局函数通过`REGISTER_LAYER`自动生成。
+* 对于每个Layer实现的时候，实现两个静态(`static`)函数，分别实现这两个函数。
+* 对于获得LayerOutputType的函数,同时完成**神经网络推导**过程。即在运行时设置ParameterSize，动态添加Layer的辅助输入等等。
+
+举例来说，例如对于FCLayer，可能的实现为:
+
+```C++
+
+class FCLayer :public Layer {
+public:
+  void init() { ... }
+  void forward() { ... }
+  void backward() { ... }
+  
+  static void getLayerDefinition(LayerDef& def) {
+    LayerDefinition::supportSize(def);
+    LayerDefinition::supportDropout(def);
+    LayerDefinition::addInput()
+        .setRepeatable(True)
+        .addSupport({ InputType::Dense, InputType::SparseInt, InputType::Sparse })
+        .addSupportSeqLevel({0, 1, 2})
+        .addDoc("FC Layer is fully connected. Blah blah blah...");
+  }
+  
+  static std::vector<LayerOutputType> getLayerOutputType(const std::vector<LayerOutputDef>& inputs,
```

Comment:
need to check whether inputs are consistent (e.g., whether dimensions and types match)